File: example_workflows/demo_account_setup/basic_logging.py
#!/usr/bin/env python3
"""
Basic Logging Example

This example demonstrates the basic logging functionality as shown in the
KeywordsAI quickstart guide.

Quickstart guide: https://docs.keywordsai.co/get-started/quickstart/logging
"""

import os
import requests
import json
from typing import Dict, List, Any, Optional
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

BASE_URL = os.getenv("KEYWORDSAI_BASE_URL", "https://api.keywordsai.co/api")
API_KEY = os.getenv("KEYWORDSAI_API_KEY")

# Model configuration from environment variables
DEFAULT_MODEL = os.getenv("DEFAULT_MODEL", "gpt-4o")
DEFAULT_MODEL_MINI = os.getenv("DEFAULT_MODEL_MINI", "gpt-4o-mini")
DEFAULT_MODEL_CLAUDE = os.getenv("DEFAULT_MODEL_CLAUDE", "claude-3-5-sonnet-20241022")


def create_log(
    model: str,
    input_messages: List[Dict[str, str]],
    output_message: Dict[str, str],
    custom_identifier: Optional[str] = None,
    span_name: Optional[str] = None,
    **kwargs
) -> Dict[str, Any]:
    """
    Create a new log entry in Keywords AI.
    
    Args:
        model: Model name (e.g., "gpt-4o", "claude-3-5-sonnet")
        input_messages: List of input messages in OpenAI format
        output_message: Output message in OpenAI format
        custom_identifier: Optional custom identifier for the log
        span_name: Optional span name for the log
        **kwargs: Additional fields to include in the log
    
    Returns:
        Dict containing the created log data
    """
    url = f"{BASE_URL}/request-logs/create"
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {API_KEY}"
    }
    
    payload = {
        "model": model,
        "input": input_messages,
        "output": output_message
    }
    
    if custom_identifier:
        payload["custom_identifier"] = custom_identifier
    if span_name:
        payload["span_name"] = span_name
    
    # Add any additional fields
    payload.update(kwargs)
    
    print("Creating log entry...")
    print(f"  URL: {url}")
    print(f"  Model: {model}")
    print(f"  Input messages: {len(input_messages)}")
    if custom_identifier:
        print(f"  Custom identifier: {custom_identifier}")
    if span_name:
        print(f"  Span name: {span_name}")
    print(f"  Request Body: {json.dumps(payload, indent=2)}")
    
    response = requests.post(url, headers=headers, json=payload)
    response.raise_for_status()
    
    data = response.json()
    print(f"\n✓ Log created successfully")
    if 'unique_id' in data:
        print(f"  Log ID (unique_id): {data.get('unique_id')}")
    if 'id' in data:
        print(f"  Log ID: {data.get('id')}")
    if 'trace_id' in data:
        print(f"  Trace ID: {data.get('trace_id')}")
    
    return data


def main():
    """Example usage of basic logging."""
    print("=" * 80)
    print("KeywordsAI Basic Logging Example")
    print("=" * 80)
    
    # Example 1: Simple log entry
    print("\n📝 Example 1: Creating a simple log entry")
    print("-" * 80)
    
    log_data = create_log(
        model=DEFAULT_MODEL,
        input_messages=[
            {"role": "user", "content": "What is the capital of France?"}
        ],
        output_message={
            "role": "assistant",
            "content": "The capital of France is Paris."
        }
    )
    
    # Example 2: Log with custom identifier
    print("\n📝 Example 2: Creating a log with custom identifier")
    print("-" * 80)
    
    log_data_2 = create_log(
        model=DEFAULT_MODEL_MINI,
        input_messages=[
            {"role": "user", "content": "Tell me a fun fact about space"}
        ],
        output_message={
            "role": "assistant",
            "content": "A day on Venus is longer than its year! Venus rotates so slowly that it takes longer to complete one rotation than to orbit the Sun once."
        },
        custom_identifier="space_fact_query_001"
    )
    
    # Example 3: Log with span name
    print("\n📝 Example 3: Creating a log with span name")
    print("-" * 80)
    
    log_data_3 = create_log(
        model=DEFAULT_MODEL_CLAUDE,
        input_messages=[
            {"role": "user", "content": "Explain quantum computing in simple terms"}
        ],
        output_message={
            "role": "assistant",
            "content": "Quantum computing uses quantum mechanical phenomena like superposition and entanglement to perform computations. Unlike classical bits that are either 0 or 1, quantum bits (qubits) can exist in multiple states simultaneously, allowing for parallel processing of information."
        },
        span_name="quantum_explanation"
    )
    
    # Example 4: Multi-turn conversation
    print("\n📝 Example 4: Creating a log for multi-turn conversation")
    print("-" * 80)
    
    log_data_4 = create_log(
        model=DEFAULT_MODEL,
        input_messages=[
            {"role": "user", "content": "What's the weather like?"},
            {"role": "assistant", "content": "I don't have access to real-time weather data. Could you tell me your location?"},
            {"role": "user", "content": "San Francisco"}
        ],
        output_message={
            "role": "assistant",
            "content": "I still can't access real-time weather, but I'd recommend checking a weather app or website like Weather.com for current conditions in San Francisco."
        },
        custom_identifier="weather_conversation_001",
        span_name="weather_assistant"
    )
    
    print("\n" + "=" * 80)
    print("All examples completed successfully!")
    print("=" * 80)
    
    return {
        "log_1": log_data,
        "log_2": log_data_2,
        "log_3": log_data_3,
        "log_4": log_data_4
    }


if __name__ == "__main__":
    main()

File: example_workflows/demo_account_setup_TS/basic_logging.ts
/**
 * Basic Logging Example
 * Documentation: https://docs.keywordsai.co/get-started/quickstart/logging
 */

import "dotenv/config";

const BASE_URL = process.env.KEYWORDSAI_BASE_URL || "https://api.keywordsai.co/api";
const API_KEY = process.env.KEYWORDSAI_API_KEY;
const DEFAULT_MODEL = process.env.DEFAULT_MODEL || "gpt-4o";
const DEFAULT_MODEL_MINI = process.env.DEFAULT_MODEL_MINI || "gpt-4o-mini";
const DEFAULT_MODEL_CLAUDE = process.env.DEFAULT_MODEL_CLAUDE || "claude-3-5-sonnet-20241022";

interface Message {
  role: string;
  content: string;
}

interface CreateLogOptions {
  model: string;
  inputMessages: Message[];
  outputMessage: Message;
  customIdentifier?: string;
  spanName?: string;
  [key: string]: unknown;
}

interface LogResponse {
  id?: string;
  unique_id?: string;
  trace_id?: string;
  [key: string]: unknown;
}

export async function createLog(options: CreateLogOptions): Promise<LogResponse> {
  const { model, inputMessages, outputMessage, customIdentifier, spanName, ...kwargs } = options;

  const url = `${BASE_URL}/request-logs/create`;
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${API_KEY}`,
  };

  const payload: Record<string, unknown> = { model, input: inputMessages, output: outputMessage };
  if (customIdentifier) payload.custom_identifier = customIdentifier;
  if (spanName) payload.span_name = spanName;
  Object.assign(payload, kwargs);

  const response = await fetch(url, { method: "POST", headers, body: JSON.stringify(payload) });
  if (!response.ok) throw new Error(`HTTP error! status: ${response.status}`);

  const data: LogResponse = await response.json();
  console.log(`[OK] Log created: ${data.unique_id || data.id}`);
  return data;
}

async function main() {
  console.log("Basic Logging Example\n");

  // Example 1: Simple log
  console.log("[1] Simple log");
  await createLog({
    model: DEFAULT_MODEL,
    inputMessages: [{ role: "user", content: "What is the capital of France?" }],
    outputMessage: { role: "assistant", content: "The capital of France is Paris." },
  });

  // Example 2: Log with custom identifier
  console.log("\n[2] Log with custom identifier");
  await createLog({
    model: DEFAULT_MODEL_MINI,
    inputMessages: [{ role: "user", content: "Tell me a fun fact about space" }],
    outputMessage: {
      role: "assistant",
      content: "A day on Venus is longer than its year!",
    },
    customIdentifier: "space_fact_query_001",
  });

  // Example 3: Log with span name
  console.log("\n[3] Log with span name");
  await createLog({
    model: DEFAULT_MODEL_CLAUDE,
    inputMessages: [{ role: "user", content: "Explain quantum computing in simple terms" }],
    outputMessage: {
      role: "assistant",
      content: "Quantum computing uses quantum mechanical phenomena to perform computations.",
    },
    spanName: "quantum_explanation",
  });

  // Example 4: Multi-turn conversation
  console.log("\n[4] Multi-turn conversation");
  await createLog({
    model: DEFAULT_MODEL,
    inputMessages: [
      { role: "user", content: "What's the weather like?" },
      { role: "assistant", content: "I don't have access to real-time weather data." },
      { role: "user", content: "San Francisco" },
    ],
    outputMessage: {
      role: "assistant",
      content: "I'd recommend checking a weather app for San Francisco.",
    },
    customIdentifier: "weather_conversation_001",
    spanName: "weather_assistant",
  });

  console.log("\nDone.");
}

const isMainModule = import.meta.url === `file://${process.argv[1]}`;
if (isMainModule) {
  main().catch(console.error);
}
